feat: support qiskit 2.0+

CalMacCQ opened the pull request:
# Description

Building off @cqc-alec 's `ae/qiskit-2` branch https://github.com/CQCL/pytket-qiskit/tree/ae/qiskit-2

Mostly requries updating how conditional logic is handled in the converters.

I've updated some of the tests where neceasary whilst preserving the intent of the test. Theres one expection to this... I have removed the conversion of `CircBox`es containing conditonal gates in `tk_to_qiskit` and updated the test to expect a `NotImplementedError` for now.



# Related issues

closes #464, #427 

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

CalMacCQ commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -860,39 +873,49 @@ def append_tk_command_to_qiskit(
+                qcirc.append(qiskit_gates.YGate().control(len(qargs) - 1), qargs)
+        return qcirc
     if optype == OpType.CnZ:
         if len(qargs) == 2:
```
TODO check that I didn't mess up the handling of controlled-Z gates in the merge -> 23289c8f4b9e7e596e6753a2eb4c528b6e3223ae

CalMacCQ commented on tests/backend_test.py:
```diff
@@ -1528,15 +1528,15 @@ def test_optimisation_level_3_compilation() -> None:
-    assert compiled_3_timeout.depth() == 132
+    assert compiled_3_timeout.n_2qb_gates() <= 81
+    assert compiled_3_timeout.n_gates <= 190
+    assert compiled_3_timeout.depth() <= 138
```
Annoyingly the C.I. tests have different numbers

https://github.com/CQCL/pytket-qiskit/actions/runs/14447570549/job/40511846182#step:10:879

CalMacCQ commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -755,6 +735,7 @@ def append_tk_command_to_qiskit(
     symb_map: dict[Parameter, sympy.Symbol],
     range_preds: dict[Bit, tuple[list["UnitID"], int]],
+    condition: tuple[ClassicalRegister | Clbit, int] | None = None,
 ) -> InstructionSet:
```
TODO check that this return type of `InstructionSet` is still correct. I think it just returns a `QuantumCircuit` now.

CalMacCQ replied:
mypy doesn't seem to complain?

CalMacCQ commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -662,29 +626,31 @@ def add_qiskit_data(
+                    optype,
+                    params,
+                    qubits + bits,
+                )  # type: ignore
```
wonder if these `# type: ignore` statements can be removed now

CalMacCQ commented on tests/qiskit_convert_test.py:
```diff
@@ -436,19 +436,10 @@ def test_conditions() -> None:
-    assert len(c1.get_commands()) == 3
-    # conversion loses rangepredicates so equality comparison not valid
+    # Converting a CircBox containing conditional gates gives an error
+    # TODO consider removing this restriction
```
Here I've decided to drop support for converting `CircBox`es which contain conditional gates in `tk_to_qiskit`. We should consdier updating this

CalMacCQ commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -747,6 +713,24 @@ def _get_params(
     return [param_to_qiskit(p, symb_map) for p in op.params]
 
 
+def _apply_qiskit_instruction(
```
Compare the use of this helper function with the logic in the [original commit](https://github.com/CQCL/pytket-qiskit/pull/475/commits/1929e60be3e076029c099c55d874acb7a36e9567)

CalMacCQ commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -747,6 +713,24 @@ def _get_params(
+def _apply_qiskit_instruction(
+    qcirc: QuantumCircuit,
+    instruc: Instruction,
+    qargs: list[UnitType.qubit],  # type: ignore
```
Is this annotation correct?

CalMacCQ commented on pytket/extensions/qiskit/backends/ibm_utils.py:
```diff
@@ -173,6 +173,9 @@ def lightsabre(
         c.rename_units({q: Node(q.index[0]) for q in c.qubits})
+        # Decompose CircBoxes corresponding to "If" and "Else" blocks in
+        #  conditional gates (qiskit IfElseOp).
+        DecomposeBoxes(excluded_types={OpType.BRIDGE}).apply(c)
```
This resolves the remote test failure here -> https://github.com/CQCL/pytket-qiskit/actions/runs/14597476884/job/40946972788

cqc-alec replied:
`BRIDGE` isn't a box type, so I don't see how this can have any effect.

CalMacCQ replied:
I'm aware of that but was confused by the error message. I've removed for now. See [432c707](https://github.com/CQCL/pytket-qiskit/pull/475/commits/432c707dce076adc6c2b16f7ca6b0bf0c86e90d4)

CalMacCQ commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -783,26 +770,52 @@ def append_tk_command_to_qiskit(
+                )
+            else:
+                instruc = subqc.to_instruction()
+                _apply_qiskit_instruction(qcirc, instruc, qargs, condition)  # type: ignore
```
Would be nice to remove the `# type: ignore` here and on line 773

Not sure why mypy doesn't like the type of the `condition` arg to `_apply_qiskit_instruction`... It should be the same type as the input to `append_tk_command_to_qiskit`.

```
Argument 4 to "_apply_qiskit_instruction" has incompatible type "tuple[Any | Any, int] | None"; expected "Iterable[Any]"  [arg-type]
```

cqc-alec replied:
Argument 4 is `cargs`. This seems to be missing from the call here (and above). Better to use named arguments to avoid this kind of bug.

CalMacCQ commented on pytket/extensions/qiskit/result_convert.py:
```diff
@@ -93,17 +93,18 @@ def qiskit_experimentresult_to_backendresult(
 
     c_bits, q_bits = None, None
-    if hasattr(header, "creg_sizes"):
+    if "creg_sizes" in header:
```
header is now a `dict` with `creg_sizes` and `memory_slots` as keys

cqc-melf commented on docs/changelog.md:
```diff
@@ -7,11 +7,14 @@
 ## Unreleased
 
 - Add support for Python 3.13.
+- Support qiskit >=2.0
```
Maybe put in a bit more clear that we are dropping support for qiskit 1.x?

CalMacCQ replied:
Thanks. See [3864d5f](https://github.com/CQCL/pytket-qiskit/pull/475/commits/3864d5f39673df8a4be328eebb2b645956b6a19a).

cqc-melf commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -747,6 +713,24 @@ def _get_params(
+            qcirc.append(instruc, qargs, cargs)
+
+
+def _has_control_flow(qc: QuantumCircuit) -> bool:
```
I think that would be good. At least put in an error message if they are not yet implemented.

CalMacCQ replied:
The reason I didn't check for those is that these operations are not handled by any of the pytket -> qiskit conversion code (there isn't a nice way to express these in pytket). Therefore a qiskit `QuantumCircuit` representing the subcircuit within a `CircBox` will not contain such operations.

Maybe the thing to do would be to rename this function to `has_if_else`. No pytket circuit can be constructed that would be converted to anything containing a `SwitchCaseOp`. 

As for the `qiskit -> pytket` direction... Consider the following example with `SwitchCaseOp`

```python
from qiskit.circuit import QuantumRegister, ClassicalRegister, QuantumCircuit

qubits = QuantumRegister(1)
clbits = ClassicalRegister(1)
circuit = QuantumCircuit(qubits, clbits)
(q0,) = qubits
(c0,) = clbits
 
circuit.h(q0)
circuit.measure(q0, c0)
with circuit.switch(c0) as case:
    with case(0):
        circuit.x(q0)
    with case(1):
        circuit.z(q0)
circuit.measure(q0, c0)
```

If we try to convert this to pytket we can an error which is hopefully helpful enough
```python
from pytket.extensions.qiskit import qiskit_to_tk

qiskit_to_tk(circuit)
```
The error

```
NotImplementedError: Conversion of qiskit's switch_case instruction is currently unsupported by qiskit_to_tk. Consider using QuantumCircuit.decompose() before attempting conversion.
```

CalMacCQ replied:
Maybe the recommedation in the error message to use the `decompose` method is a little dated as this would be more helpful for higher level unitary operations.

CalMacCQ replied:
Personally I think representing anything more complited than `IfElseOp` is a job for HUGR rather than pytket.

cqc-melf commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -662,29 +626,31 @@ def add_qiskit_data(
             elif optype == OpType.CircBox:
                 circbox = _build_circbox(instr, circuit)
-                self.tkc.add_circbox(circbox, qubits + bits, **condition_kwargs)  # type: ignore
+                self.tkc.add_circbox(circbox, qubits + bits)  # type: ignore
```
It is not really clear to me where this condition has gone? Can you give maybe some details?

cqc-melf replied:
(Same for the two below)

CalMacCQ replied:
These `condition_kwargs` dictionaries were populated from the `Instruction.condition` property. In qiskit v2 this property is removed. Conditional gates in qiskit v2 are `IfElseOp` instances which have a `condition` property. These `IfElseOp` instructions have their own handling which was added in previous PRs.

CalMacCQ commented:
Two points that occured to me

1. I think I missed the use of `_apply_qiskit_instruction` helper for controlled-Z gates
2. I think its fair to say we've expanded the range of conditional ops which can be converted from pytket -> qiskit. Therefore it probably makes sense to have some more tests.

cqc-alec commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -910,25 +943,32 @@ def append_tk_command_to_qiskit(
     if optype == OpType.Phase:
         params = _get_params(op, symb_map)
         assert len(params) == 1
+        # TODO is there a way to make this conditional?
```
If not we should maybe print a warning message if there is a condition here.

cqc-alec commented on tests/backend_test.py:
```diff
@@ -1528,15 +1528,15 @@ def test_optimisation_level_3_compilation() -> None:
-    assert compiled_2.n_2qb_gates() == 68
-    assert compiled_2.n_gates == 186
-    assert compiled_2.depth() == 130
+    assert compiled_2.n_2qb_gates() <= 69
```
This is the one that most concerns me. Previously this circuit compiled to 68 2-qubit gates at optimization level 2; now it compiles to 69. The circuit contains no conditional gates so it's hard to see where this could be coming from...

cqc-alec replied:
I've experimented with changing the parameters of the circuit (numbers of iterations of the two loops), and it seems that performance is sometimes worse, sometimes the same, sometimes better. I think we can probably put this down to different behaviour of `lightsabre` in qiskit 2, so happy to accept these numbers.

yao-cqc commented:
```python
from pytket.circuit import Circuit
from pytket.extensions.qiskit.qiskit_convert import qiskit_to_tk, tk_to_qiskit
from qiskit.transpiler.preset_passmanagers.level3 import level_3_pass_manager
from qiskit.transpiler import CouplingMap, PassManagerConfig
import json

circuit_json = '{"bits": [["c0", [0]], ["c1", [0]], ["c2", [0]], ["c3", [0]]], "commands": [{"args": [["q", [0]]], "op": {"type": "H"}}, {"args": [["q", [1]]], "op": {"type": "H"}}, {"args": [["q", [2]]], "op": {"type": "H"}}, {"args": [["q", [3]]], "op": {"type": "H"}}, {"args": [["q", [0]], ["q", [1]], ["q", [2]], ["q", [3]]], "op": {"data": "", "signature": ["Q", "Q", "Q", "Q"], "type": "Barrier"}}, {"args": [["q", [0]]], "op": {"type": "H"}}, {"args": [["q", [0]], ["c0", [0]]], "op": {"type": "Measure"}}, {"args": [["c0", [0]], ["q", [1]]], "op": {"conditional": {"op": {"params": ["1/2"], "type": "U1"}, "value": 1, "width": 1}, "type": "Conditional"}}, {"args": [["c0", [0]], ["q", [2]]], "op": {"conditional": {"op": {"params": ["1/4"], "type": "U1"}, "value": 1, "width": 1}, "type": "Conditional"}}, {"args": [["c0", [0]], ["q", [3]]], "op": {"conditional": {"op": {"params": ["1/8"], "type": "U1"}, "value": 1, "width": 1}, "type": "Conditional"}}, {"args": [["q", [1]]], "op": {"type": "H"}}, {"args": [["q", [1]], ["c1", [0]]], "op": {"type": "Measure"}}, {"args": [["c1", [0]], ["q", [2]]], "op": {"conditional": {"op": {"params": ["1/2"], "type": "U1"}, "value": 1, "width": 1}, "type": "Conditional"}}, {"args": [["c1", [0]], ["q", [3]]], "op": {"conditional": {"op": {"params": ["1/4"], "type": "U1"}, "value": 1, "width": 1}, "type": "Conditional"}}, {"args": [["q", [2]]], "op": {"type": "H"}}, {"args": [["q", [2]], ["c2", [0]]], "op": {"type": "Measure"}}, {"args": [["c2", [0]], ["q", [3]]], "op": {"conditional": {"op": {"params": ["1/2"], "type": "U1"}, "value": 1, "width": 1}, "type": "Conditional"}}, {"args": [["q", [3]]], "op": {"type": "H"}}, {"args": [["q", [3]], ["c3", [0]]], "op": {"type": "Measure"}}], "created_qubits": [], "discarded_qubits": [], "implicit_permutation": [[["q", [0]], ["q", [0]]], [["q", [1]], ["q", [1]]], [["q", [2]], ["q", [2]]], [["q", [3]], ["q", [3]]]], "name": "inverseqft2", "phase": "0.0", "qubits": [["q", [0]], ["q", [1]], ["q", [2]], ["q", [3]]]}'

circ = Circuit.from_dict(json.loads(circuit_json))
qc = tk_to_qiskit(circ)

coupling = CouplingMap([[0,1],[1,0],[1,2],[2,1],[2,3],[3,2],[3,4],[4,3]])
config = PassManagerConfig(
    coupling_map=coupling,
    basis_gates=["cx", "sx", "x", "rz", "if_else"],
    seed_transpiler=0,
)
pass_manager = level_3_pass_manager(config)
compiled_qc = pass_manager.run(qc)
tk_circ = qiskit_to_tk(compiled_qc)
```
raises
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[35], [line 20](vscode-notebook-cell:?execution_count=35&line=20)
     [18](vscode-notebook-cell:?execution_count=35&line=18) pass_manager = level_3_pass_manager(config)
     [19](vscode-notebook-cell:?execution_count=35&line=19) compiled_qc = pass_manager.run(qc)
---> [20](vscode-notebook-cell:?execution_count=35&line=20) tk_circ = qiskit_to_tk(compiled_qc)
...
File ~/.pyenv/versions/pytket-benchmarking-new/lib/python3.11/site-packages/pytket/extensions/qiskit/qiskit_convert.py:581, in <listcomp>(.0)
    [578](https://file+.vscode-resource.vscode-cdn.net/Users/yaotang/Desktop/cqc_dev/pytket-benchmarking/~/.pyenv/versions/pytket-benchmarking-new/lib/python3.11/site-packages/pytket/extensions/qiskit/qiskit_convert.py:578) for datum in data:
    [579](https://file+.vscode-resource.vscode-cdn.net/Users/yaotang/Desktop/cqc_dev/pytket-benchmarking/~/.pyenv/versions/pytket-benchmarking-new/lib/python3.11/site-packages/pytket/extensions/qiskit/qiskit_convert.py:579)     instr, qargs, cargs = datum.operation, datum.qubits, datum.clbits
--> [581](https://file+.vscode-resource.vscode-cdn.net/Users/yaotang/Desktop/cqc_dev/pytket-benchmarking/~/.pyenv/versions/pytket-benchmarking-new/lib/python3.11/site-packages/pytket/extensions/qiskit/qiskit_convert.py:581)     qubits: list[Qubit] = [self.qbmap[qbit] for qbit in qargs]
    [582](https://file+.vscode-resource.vscode-cdn.net/Users/yaotang/Desktop/cqc_dev/pytket-benchmarking/~/.pyenv/versions/pytket-benchmarking-new/lib/python3.11/site-packages/pytket/extensions/qiskit/qiskit_convert.py:582)     bits: list[Bit] = [self.cbmap[bit] for bit in cargs]
    [584](https://file+.vscode-resource.vscode-cdn.net/Users/yaotang/Desktop/cqc_dev/pytket-benchmarking/~/.pyenv/versions/pytket-benchmarking-new/lib/python3.11/site-packages/pytket/extensions/qiskit/qiskit_convert.py:584)     optype = None

KeyError: <Qubit register=(4, "q"), index=1>
```

CalMacCQ commented:
>     1. I think I missed the use of `_apply_qiskit_instruction` helper for controlled-Z gates

fixed ->  [bf293b4](https://github.com/CQCL/pytket-qiskit/pull/475/commits/bf293b4f8553d778bc5d15d279919ed85c51f2de)

>     2. I think its fair to say we've expanded the range of conditional ops which can be converted from pytket -> qiskit. Therefore it probably makes sense to have some more tests.

Added a test that the condtional ops generated by `tk_to_qiskit` are right -> [aa188b4](https://github.com/CQCL/pytket-qiskit/pull/475/commits/aa188b4e971cf314ceef7fb93344ec9465d7dfbf)

cqc-alec commented:
I think this PR also closes #427 .

cqc-alec commented:
Sorry, closed by mistake.

CalMacCQ commented:
@yao-cqc

I think the issue with [this commit ]([e7b7f77](https://github.com/CQCL/pytket-qiskit/pull/475/commits/e7b7f7706c390745af41d49741f6654cea562901)) which caused the [test failures](https://github.com/CQCL/pytket-qiskit/actions/runs/14623121203/job/41028002720?pr=475) is that the gates in the interior of the `IfElseOp` object are getting the qubit args (`true_qargs`) of the entire if-else subroutine.... So for example the x gate [here](https://github.com/CQCL/pytket-qiskit/blob/main/tests/qiskit_convert_test.py#L1358) is getting both qubits from inside the `IfElseOp` as an argument. Took me a while to find this.

The ` IndexError: vector` error message was not very enlightening.

CalMacCQ commented:
Converting circuits which are transpiled is a little tricky. I think it would be worth having a think about how to handle the `QuantumCircuit.layout` property -> https://github.com/CQCL/pytket-qiskit/issues/123 I had forgotten about this issue.

CalMacCQ commented:
> @yao-cqc
> 
> I think the issue with this commit which caused the [test failures](https://github.com/CQCL/pytket-qiskit/actions/runs/14623121203/job/41028002720?pr=475) is that the gates in the interior of the `IfElseOp` object are getting the qubit args (`true_qargs`) of the entire if-else subroutine.... So for example the x gate [here](https://github.com/CQCL/pytket-qiskit/blob/main/tests/qiskit_convert_test.py#L1358) is getting both qubits from inside the `IfElseOp` as an argument. Took me a while to find this.
> 
> The `IndexError: vector` error message was not very enlightening. Maybe I should make an issue to improve it.
> 
> To fix this error we will need to ensure that each gate within the `IfElseOp` gets the correct subset of qubits as an argument.

This wasn't quite accurate. In the tests it seems that the single qubit gates are all being added to the same qubit and then the blank wires are being removed by the call to `Circuit.remove_blank_wires` meaning that the `CircBox` is the wrong size. Still the issue is about passing the right qubit args to the gates within the `CircBox` representing the `IfElseOp`.

yao-cqc commented:
@CalMacCQ 
I have made some changes to the if_else conversion code so it now passes all tests except `test_conditional_conversion_2`, which I believe is caused by https://github.com/CQCL/tket/issues/1881

CalMacCQ commented on tests/qiskit_convert_test.py:
```diff
@@ -1419,3 +1426,79 @@ def test_nested_conditionals() -> None:
+    )
+    pass_manager = level_3_pass_manager(config)
+    compiled_qc = pass_manager.run(qc)
+    tk_circ = qiskit_to_tk(compiled_qc)
```
This converted circuit has 3 scratch bits. Does this seem right to you?

CalMacCQ commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -487,75 +460,109 @@ def _build_circbox(instr: Instruction, circuit: QuantumCircuit) -> CircBox:
+    # We rename the units to make sure the registers in the inner circuit
+    # is a subset of the registers in the ourter circuit.
+    rename_map = {}
+    for i, inner_q in enumerate(if_qc.qubits):
```
This logic is repeated for the "else block". Could make ahelper. What do you think?

CalMacCQ commented:
Fixed some mypy complaints in [79c83a1](https://github.com/CQCL/pytket-qiskit/pull/475/commits/79c83a11519f884246785d9ed67fc35df462f78d)

Still have some left to sqaush

```
mypy --config-file=mypy.ini --no-incremental -p pytket -p tests
pytket/extensions/qiskit/qiskit_convert.py:485: error: Invalid index type "Bit" for "dict[Qubit, Qubit]"; expected type "Qubit"  [index]
pytket/extensions/qiskit/qiskit_convert.py:485: error: Incompatible types in assignment (expression has type "Bit", target has type "Qubit")  [assignment]
pytket/extensions/qiskit/qiskit_convert.py:509: error: Invalid index type "Bit" for "dict[Qubit, Qubit]"; expected type "Qubit"  [index]
pytket/extensions/qiskit/qiskit_convert.py:509: error: Incompatible types in assignment (expression has type "Bit", target has type "Qubit")  [assignment]
```

cqc-alec commented on pytket/extensions/qiskit/qiskit_convert.py:
```diff
@@ -590,33 +593,47 @@ def __init__(
+                    self.cbmap[cb] = Bit(reg.name, i)
+
+    @classmethod
+    def from_bits(
```
Why is this called `from_bits()` when it takes qubits as well?

cqc-alec reviewed (commented):
Need to fix the mypy issues.

CalMacCQ commented:
> Need to fix the mypy issues.

see [ef10679](https://github.com/CQCL/pytket-qiskit/pull/475/commits/ef1067989394cd3a020272d16ece0a5ef5bee6bb)
